### ISSUE TITLE
Is the chi_p function in pycbc correct?

### ISSUE THREAD

**Issue by Cyberface** [gwastro/pycbc#2247]
I was looking at the chi_p function https://github.com/gwastro/pycbc/blob/master/pycbc/conversions.py#L395
and I don't think it is correctly implementing Eq. 3.4 of 1408.1810

Comparison between pycbc function and my function can be performed
with the following code snippet as well as my independent chi_p function below too.

My hunch is that the pycbc function seems to be just returning the magnitude
of the in-plane component on the primary black-hole but TBH the code
is a little too dense for me to read easily.

```
from pycbc.conversions import chi_p

# primary black hole parameters
m_primary = 10.
sx_p,sy_p = 0.1, 0.

# seondary black hole parameters
m_secondary = 3.
sx_s,sy_s = 0.5, 0.

print(chi_p(m_secondary, m_primary, sx_s, sy_s, sx_p, sy_p))
# this equals 0.1

# the code for 'khan_chi_p' function is below.
print(khan_chi_p(m_secondary, m_primary, sx_s, sy_s, sx_p, sy_p))
# this equals 0.1285

# evaluating both the pycbc function and my function yield different numbers.
# which is a little concerning.

```

Here is my independent implementation:
Note that my convention in this function for the masses follows that of
1408.1810. Namely m2>=m1 where q=m2/m1.

```
import numpy as np

def A1(q):
    """
    A1 defined under eq. 3.2 of 1408.1810
    """
    return 2. + 3.*q/2.

def A2(q):
    """
    A2 defined under eq. 3.2 of 1408.1810
    """
    return 2. + 3./(2.*q)

def Sp(m1, m2, s1x, s1y, s2x, s2y):
    """
    eq. 3.3 of 1408.1810
    m2 >= m1
    """
    q = m2/m1

    A1val = A1(q)
    A2val = A2(q)

    # convert dimensionless spins to angular momenta.
    # S_i = chi_i * m_i**2
    S1perp = np.sqrt(s1x**2. + s1y**2.) * m1**2
    S2perp = np.sqrt(s2x**2. + s2y**2.) * m2**2

    Spval = np.max( [[A1val*S1perp], [A2val*S2perp]] )
    return Spval

def khan_chi_p(m1, m2, s1x, s1y, s2x, s2y):
    """
    eq. 3.4 of 1408.1810
    m2 >= m1
    """
    q = m2/m1 # Note this is the convention used in 1408.1810. See first paragraph of section 3, page 4.
    Spval = Sp(m1, m2, s1x, s1y, s2x, s2y)

    A2val = A2(q)

    chip = Spval / (A2val * m2**2.)

    return chip

# primary black hole parameters
m_primary = 10.
sx_p,sy_p = 0.1, 0.

# seondary black hole parameters
m_secondary = 3.
sx_s,sy_s = 0.5, 0.

print(khan_chi_p(m_secondary, m_primary, sx_s, sy_s, sx_p, sy_p))
```

**Comment by ahnitz** [comment]
@Cyberface It looks like that's what it is doing to me too... which definitely seems wrong!

**Comment by cdcapano** [comment]
@Cyberface It's a bit hard for me to follow the equations in `conversions.py` myself to see what's going wrong, but I'm inclined to believe you since your code is a straight forward implementation of 1408.1810.

The current equations in `conversions.py` came from: https://dcc.ligo.org/DocDB/0140/T1700056/002/spin_parametrisation.pdf

I can't figure out if the issue is in our implementation, or the equations in that document, or my understanding of the equations in that document! In any event, can you fill out a PR with your implementation of chi_p ASAP?

**Comment by Cyberface** [comment]
OK I found out what was going wrong.

**Passing integers into the pycbc function `chi_p` causes a rounding error and computes chi_p incorrectly.**

My original code block was wrong. I wrote that I was passing `floats` in my code block by I *actually* was passing integers. So it should read

```
from pycbc.conversions import chi_p

# primary black hole parameters
m_primary = 10
sx_p,sy_p = 0.1, 0.

# seondary black hole parameters
m_secondary = 3
sx_s,sy_s = 0.5, 0.

print(chi_p(m_secondary, m_primary, sx_s, sy_s, sx_p, sy_p))
# this equals 0.1
```

If instead you set 

```
m_primary = 10.0
m_secondary = 3.0
```
They you get the correct value of chip which is `0.12857142`.



The current function, though hard to read agrees with LAL and my own independent version. So there is no need to change that really.

For future reference you can also double check with what LAL computes for chi_p too using.
Here is an example showing they agree.

```
import lal
import lalsimulation as lalsim

m1 = 10.
m2 = 3.
s1x,s1y,s1z = 0.1, 0., 0.
s2x,s2y,s2z = 0.5, 0., 0.

_,_,chip_lal,_,_,_,_ = lalsim.SimIMRPhenomPCalculateModelParametersFromSourceFrame(m1*lal.MSUN_SI, m2*lal.MSUN_SI, 10, 0, 0, s1x,s1y,s1z,s2x,s2y,s2z,2)

print(chip_lal)
#0.12857142857142856

from pycbc.conversions import chi_p
pycbc_chi_p = chi_p(m1,m2,s1x,s1y,s2x,s2y)
print(pycbc_chi_p)
#0.12857142857



```

**Comment by Cyberface** [comment]
The test would then depend on lalsuite. Is that ok?

**Comment by ahnitz** [comment]
Yes that's ok.

**Comment by tdent** [comment]
hang on, are we saying it's OK for a function to give wrong output when fed integer input values?

**Comment by ahnitz** [comment]
@tdent, no, one should probably import future division in his module or make sure to cast to a float. That's a patch better added by @cdcapano

**Comment by spxiwh** [comment]
Future division please!

**Comment by cdcapano** [comment]
Sorry, I misunderstood @Cyberface. I thought he said his code was the one that wasn't casting to float. Seb, can you clarify?

But yeah, if conversions isn't casting properly, I agree it'd be best to use future division.

**Comment by ahnitz** [comment]
@cdcapano you might need to set the dtype for you array conversion as well.

**Comment by Cyberface** [comment]
It's a problem with the pycbc function.

Said differently: we need to go back... to the 'from future import division'

**Comment by duncan-brown** [comment]
@Cyberface please can you close this if #2248 addresses the issue?

**Comment by Cyberface** [comment]
@duncan-brown thanks for pinging me on this, this was not fixed in #2248
